maro/cli/grass/executors/grass_on_premises_executor.py

```diff
@@ -0,0 +1,404 @@
+            "root['master']['fluentd']": {'port': 24224},
+            "root['master']['fluentd']['port']": 24224,
+            "root['master']['samba']": {'password': ''.join(secrets.choice(alphabet) for _ in range(20))},
+            "root['master']['samba']['password']": ''.join(secrets.choice(alphabet) for _ in range(20))
+        }
+        with open(os.path.expanduser(
+                f'{GlobalPaths.MARO_GRASS_LIB}/deployments/internal/grass-on-premises-create.yml')) as fr:
+            create_deployment_template = yaml.safe_load(fr)
+        validate_and_fill_dict(
+            template_dict=create_deployment_template,
+            actual_dict=create_deployment,
+            optional_key_to_value=optional_key_to_value
+        )
+
+    def create(self):
+        logger.info("Creating cluster")
+
+        # Start creating
+        try:
+            self._set_cluster_id()
+            self._set_master_info()
+            self._init_master()
+        except Exception as e:
+            # If failed, remove details folder, then raise
+            rmtree(os.path.expanduser(f"{GlobalPaths.MARO_CLUSTERS}/{self.cluster_name}"))
+            raise e
+
+        logger.info_green(f"Cluster {self.cluster_name} has been created.")
+
+    def _set_cluster_id(self):
+        # Load details
+        cluster_details = self.cluster_details
+
+        # Set cluster id
+        cluster_details["id"] = generate_cluster_id()
+
+        # Save details
+        save_cluster_details(
+            cluster_name=self.cluster_name,
+            cluster_details=cluster_details
+        )
+
+    def _set_master_info(self):
+        # Load details
+        cluster_details = self.cluster_details
+        cluster_id = cluster_details["id"]
+        master_details = cluster_details["master"]
+        # hostname = f"{cluster_id}-master-vm"
+        hostname = cluster_details["master"]["public_ip_address"]
+        master_details["private_ip_address"] = cluster_details["master"]["public_ip_address"]
+        master_details["hostname"] = hostname
+        master_details["resource_name"] = f"{cluster_id}-master-vm"
+        admin_username = cluster_details["user"]["admin_username"]
+        public_ip_address = cluster_details["master"]["public_ip_address"]
+        logger.info_green(f"You can login to your master node with: ssh {admin_username}@{public_ip_address}")
+
+    def _init_master(self):
+        logger.info("Initializing master node")
+
+        # Load details
+        cluster_details = self.cluster_details
+        master_details = cluster_details["master"]
+        admin_username = cluster_details["user"]["admin_username"]
+        master_public_ip_address = cluster_details["master"]["public_ip_address"]
+
+        # Make sure master is able to connect
+        self.grass_executor.retry_until_connected(node_ip_address=master_public_ip_address)
+
+        # Create folders
+        self.grass_executor.remote_mkdir(
+            path=GlobalPaths.MARO_GRASS_LIB,
```

Comment:
Could path to be a list of folder path, create by one call?

Reply:
Good idea, refactored.